File: desktop/src/assist/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type {
  AssistReplySuggestionsPayload,
  AssistSegmentUpdatePayload,
  TranscriptSegmentPayload,
} from "./types";
import type { AssistAction } from "./state";

const isNonBlankString = (value: unknown): value is string =>
  typeof value === "string" && value.trim().length > 0;

function transcriptPayload(value: unknown): TranscriptSegmentPayload | null {
  if (!value || typeof value !== "object") return null;
  const payload = value as Record<string, unknown>;
  return ["transcript_id", "text", "timestamp", "source", "speaker"].every((key) => isNonBlankString(payload[key]))
    ? payload as unknown as TranscriptSegmentPayload
    : null;
}

function segmentUpdatePayload(value: unknown): AssistSegmentUpdatePayload | null {
  if (!value || typeof value !== "object") return null;
  const payload = value as Record<string, unknown>;
  const validState = ["processing", "ready", "failed", "unavailable"].includes(String(payload.state));
  const validCapability = payload.capability === "translation" || payload.capability === "simplification";
  if (!isNonBlankString(payload.transcript_id) || !validState || !validCapability) return null;
  if (payload.state === "ready" && payload.capability === "translation" && !isNonBlankString(payload.translated_text)) return null;
  if (payload.state === "ready" && payload.capability === "simplification" && (!isNonBlankString(payload.simplified_text) || (payload.target_level !== "b1" && payload.target_level !== "b2"))) return null;
  return payload as unknown as AssistSegmentUpdatePayload;
}

function replySuggestionsPayload(value: unknown): AssistReplySuggestionsPayload | null {
  if (!value || typeof value !== "object") return null;
  const payload = value as Record<string, unknown>;
  if (!isNonBlankString(payload.anchor_transcript_id) || !["processing", "ready", "failed", "unavailable"].includes(String(payload.state))) return null;
  if (payload.state === "ready" && (!Array.isArray(payload.suggestions) || payload.suggestions.some((item) => !item || typeof item !== "object" || !isNonBlankString((item as Record<string, unknown>).text) || !isNonBlankString((item as Record<string, unknown>).tone)))) return null;
  return payload as unknown as AssistReplySuggestionsPayload;
}

export async function subscribeToAssistEvents(dispatch: (action: AssistAction) => void): Promise<() => void> {
  let active = true;
  const listeners = await Promise.all([
    listen("assist://transcript-segment", (event) => {
      const payload = transcriptPayload(event.payload);
      if (active && payload) dispatch({ type: "transcript", payload });
    }),
    listen("assist://segment-update", (event) => {
      const payload = segmentUpdatePayload(event.payload);
      if (active && payload) dispatch({ type: "segment-update", payload });
    }),
    listen("assist://reply-suggestions", (event) => {
      const payload = replySuggestionsPayload(event.payload);
      if (active && payload) dispatch({ type: "reply-suggestions", payload });
    }),
  ]);
  return () => {
    active = false;
    listeners.forEach((unlisten: UnlistenFn) => unlisten());
  };
}

File: desktop/src-tauri/src/assist_mode/events.rs
//! Typed, privacy-safe desktop events emitted from backend Assist Mode messages.

use serde::Serialize;
use tauri::{AppHandle, Emitter};
use thiserror::Error;
use uuid::Uuid;

pub(crate) const TRANSCRIPT_SEGMENT_EVENT: &str = "assist://transcript-segment";
pub(crate) const SEGMENT_UPDATE_EVENT: &str = "assist://segment-update";
pub(crate) const REPLY_SUGGESTIONS_EVENT: &str = "assist://reply-suggestions";

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub(crate) struct TranscriptSegmentEvent {
    pub(crate) transcript_id: Uuid,
    pub(crate) text: String,
    pub(crate) timestamp: String,
    pub(crate) source: String,
    pub(crate) speaker: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub(crate) enum AssistSegmentCapability {
    Translation,
    Simplification,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub(crate) enum AssistUpdateState {
    Processing,
    Ready,
    Failed,
    Unavailable,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub(crate) struct AssistSegmentUpdateEvent {
    pub(crate) transcript_id: Uuid,
    pub(crate) capability: AssistSegmentCapability,
    pub(crate) state: AssistUpdateState,
    pub(crate) translated_text: Option<String>,
    pub(crate) simplified_text: Option<String>,
    pub(crate) target_level: Option<String>,
    pub(crate) message: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub(crate) struct ReplySuggestionEvent {
    pub(crate) text: String,
    pub(crate) tone: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub(crate) struct AssistReplySuggestionsEvent {
    pub(crate) anchor_transcript_id: Uuid,
    pub(crate) state: AssistUpdateState,
    pub(crate) suggestions: Vec<ReplySuggestionEvent>,
    pub(crate) message: Option<String>,
}

#[derive(Debug, Error)]
#[error("Assist Mode event delivery failed.")]
pub(crate) struct AssistEventDeliveryError;

/// A narrow, fakeable boundary for typed desktop event delivery.
pub(crate) trait AssistEventSink: Send + Sync + 'static {
    fn emit_transcript_segment(
        &self,
        event: TranscriptSegmentEvent,
    ) -> Result<(), AssistEventDeliveryError>;
    fn emit_segment_update(
        &self,
        event: AssistSegmentUpdateEvent,
    ) -> Result<(), AssistEventDeliveryError>;
    fn emit_reply_suggestions(
        &self,
        event: AssistReplySuggestionsEvent,
    ) -> Result<(), AssistEventDeliveryError>;
}

/// Tauri-local event sink. Emission failures intentionally expose no detail.
pub(crate) struct TauriAssistEventSink {
    app_handle: AppHandle,
}

impl TauriAssistEventSink {
    pub(crate) fn new(app_handle: AppHandle) -> Self {
        Self { app_handle }
    }
}

impl AssistEventSink for TauriAssistEventSink {
    fn emit_transcript_segment(
        &self,
        event: TranscriptSegmentEvent,
    ) -> Result<(), AssistEventDeliveryError> {
        self.app_handle
            .emit(TRANSCRIPT_SEGMENT_EVENT, event)
            .map_err(|_| AssistEventDeliveryError)
    }

    fn emit_segment_update(
        &self,
        event: AssistSegmentUpdateEvent,
    ) -> Result<(), AssistEventDeliveryError> {
        #[cfg(debug_assertions)]
        let capability = match event.capability {
            AssistSegmentCapability::Translation => "translation",
            AssistSegmentCapability::Simplification => "simplification",
        };
        self.app_handle
            .emit(SEGMENT_UPDATE_EVENT, event)
            .map_err(|_| AssistEventDeliveryError)?;
        #[cfg(debug_assertions)]
        eprintln!("assist event emitted capability={capability}");
        Ok(())
    }

    fn emit_reply_suggestions(
        &self,
        event: AssistReplySuggestionsEvent,
    ) -> Result<(), AssistEventDeliveryError> {
        self.app_handle
            .emit(REPLY_SUGGESTIONS_EVENT, event)
            .map_err(|_| AssistEventDeliveryError)
    }
}
